### PR TITLE
Initial Support for Windows Based Docker Images

### PR DESCRIPTION
**Pull request by AndrewKahr**
#### Changes

- Added dockerfiles for windows-based containers
- Added workflow files to build windows-based base, hub, and editor images, mirroring the existing system in place for ubuntu based images
- Modified pull-request workflow file to also test windows images

#### Checklist

- [x] Read the contribution [guide](../CONTRIBUTING.md) and accept the [code](../CODE_OF_CONDUCT.md) of conduct
- [x] Readme (updated or not needed)

### CODE REVIEW

File: images/windows/editor/100-characters-dockerhub-description.txt [hunk 1]
```diff
@@ -0,0 +1 @@
+Dockerised windows-based Unity Editor.
```
Comment (webbertakken) [inline images/windows/editor/100-characters-dockerhub-description.txt:1]: can be removed

File: images/windows/editor/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+ARG version
+ARG changeSet
+RUN "C:/Program Files/Unity Hub/Unity Hub.exe" -- --headless install \
+                                               --version %version% \
+                                               --changeset %changeSet% \
+                                               --module windows-il2cpp \
+                                               & if %ERRORLEVEL% EQU 1 exit 0
+
+# TODO: Handle additional modules
+
```
Comment (webbertakken) [inline images/windows/editor/Dockerfile:22]: module should be passed in through the workflow (matrix in new_build flow or retry args in retry flow)

File: images/windows/hub/100-characters-dockerhub-description.txt [hunk 3]
```diff
@@ -0,0 +1 @@
+Dockerised windows-based Unity Hub.
```
Comment (webbertakken) [inline images/windows/hub/100-characters-dockerhub-description.txt:1]: can be removed

**Review by webbertakken — CHANGES_REQUESTED** [review]
Amazing work! ❤

Looking good for the most part, a things still need changing to get it into a mergeable state.

File: .github/workflows/new-windows-post-2019-2-editor-image-requested.yml [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+            )
+            $Image = $Repository+":"+$Tag
+            docker manifest inspect $Image | Out-Null
+            return $?
+          }
+
+          if( (Docker-Tag-Exists -Repository unityci/editor -Tag windows-${{ github.event.client_payload.editorVersion }}-${{ matrix.targetPlatform }}-${{ github.event.client_payload.repoVersionFull }}) ) {
+            echo "Image already exists. Exiting."
+            exit 1
+          }
```
[the remark this answers is not in the document]
Comment (AndrewKahr) [reply .github/workflows/new-windows-post-2019-2-editor-image-requested.yml:77]: I've tested it locally, and it does indeed exit with the error code 1 which should trigger github actions to see an error and exit. Is this the behavior that is needed and does this need testing on a github run?

File: .github/workflows/new-windows-post-2019-2-editor-image-requested.yml [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+      
+      #######################
+      #   Free disk space   #
+      #######################
+      # TODO: Determine what we can free on windows
```
[the remark this answers is not in the document]
Comment (AndrewKahr) [reply .github/workflows/new-windows-post-2019-2-editor-image-requested.yml:82]: Can you elaborate on what it means to need the files on the host system + the built result?

File: .github/workflows/pull-requests.yml [hunk 6]
```diff
@@ -211,6 +211,87 @@ jobs:
...
+      #   Editor image   #
+      ####################
+      - name: Build Windows Editor Image # and push
+        id: build_windows_editor_image
+        run: |
+              docker build ./images/windows/editor/ `
+                --build-arg baseImage=localhost:5000/unityci/base:windows-latest
+                --build-arg hubImage=localhost:5000/unityci/hub:windows-latest
+                --build-arg version=2020.1.4f1
+                --build-arg changeSet=fa717bb873ec
```
[the remark this answers is not in the document]
Comment (AndrewKahr) [reply .github/workflows/pull-requests.yml:280]: This is part of the PR workflow, I'm not sure the versioning backend has a say in this though maybe I am mistaken on how this is triggered. Also the Ubuntu one hardcodes the version and changeSet in the same way. Should these both use versioning backend args?

File: .github/workflows/pull-requests.yml [hunk 7]
```diff
@@ -211,6 +211,87 @@ jobs:
...
+    services:
+      registry:
+        image: registry:2
+        ports:
+          - 5000:5000
+    strategy:
+      fail-fast: false
+      matrix:
+        targetPlatform:
+          - windows-il2cpp
```
[the remark this answers is not in the document]
Comment (AndrewKahr) [reply .github/workflows/pull-requests.yml:226]: https://support.unity.com/hc/en-us/articles/115005327866-How-do-I-develop-to-Xbox-platforms-
It doesn't appear that we will be able to support Xbox due to licensing. I can take a look into whether we can add in UWP support though

[the remark this answers is not in the document]
Comment (AndrewKahr) [reply .github/workflows/pull-requests.yml:226]: Looking at UWP, it requires an additional VS module for UWP build support, the GitHub runners already have it installed, not sure if anything is needed on the docker side to make it work (ie reg keys to point to the module). I think this is out of scope for this initial iteration which is mainly focused on il2cpp, however, it shouldn't be too difficult to add support for it down the line with the structure that is in place.

File: .github/workflows/pull-requests.yml (hunk 6 above)
[the remark this answers is not in the document]
Comment (webbertakken) [reply .github/workflows/pull-requests.yml:278]: No, you're right this part is supposed to be hardcoded.

**Review by webbertakken — APPROVED** [review]

**Pull request merged**